### PR TITLE
JDBC generic interpreter

### PR DESCRIPTION
**Pull request by vgmartinez**
Only you need to add to the classpath the jdbc connector jar and the interpreter add the particular properties for each db.
In the file zeppelin-daemon.sh add:
ZEPPELIN_CLASSPATH+=":${ZEPPELIN_HOME}/jdbc/jdbc/connector jar"

### CODE REVIEW

**Comment by felixcheung** [comment]
@tzolov

File: jdbc/pom.xml [hunk 1]
```diff
@@ -44,7 +44,13 @@
       <version>${project.version}</version>
       <scope>provided</scope>
     </dependency>
-
+	
+	<dependency>
```
Comment (jongyoul) [inline jdbc/pom.xml]: @vgmartinez I think this dependency doesn't need any longer. Could you please remove this?

Comment (jongyoul) [inline jdbc/pom.xml]: @vgmartinez I think this dependency doesn't need any longer. Could you please remove this?

**Comment by jongyoul** [comment]
@vgmartinez https://github.com/vgmartinez/incubator-zeppelin/blob/jdbc_interpreter/jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java#L146 How about connecting without user or password? It's better use `getConnection(String url, Properties info)`
https://docs.oracle.com/javase/8/docs/api/java/sql/DriverManager.html#getConnection-java.lang.String-java.util.Properties- ?

**Comment by jongyoul** [comment]
@vgmartinez Thanks for contribution on this issue which @Leemoonsoo and I talk about a few days ago.  It will covers several Interpreters using JDBC.

**Comment by jongyoul** [comment]
@vgmartinez Finally, How about adding a `ZEPPELIN_INTP_CLASSPATH` to zeppelin-env.sh and making it added by interpreter.sh?

**Comment by vgmartinez** [comment]
Hi @jongyoul,
Thanks for the feedback...I will implement the suggestion....

**Comment by felixcheung** [comment]
There's another earlier attempt for this #211, #60
Other JDBC-like use cases perhaps you could also factor into your design? #73, #110

**Comment by jongyoul** [comment]
@felixcheung There are several implementations for jdbc. We should talk about it for focusing one issue. Do you have any idea?

+ @Leemoonsoo

**Comment by Leemoonsoo** [comment]
As @felixcheung mentioned, there were couple of attempt.
This patch renames existing postgresql to jdbc. So if renaming does not bring big problem, I think it can be a basement and starting point to do all the things discussed in previous attempt.

What do you think? will rename postgresql to jdbc make some problem?

**Comment by felixcheung** [comment]
If I recall there were 3 areas of discussions around JDBC:
1. JDBC url, authentication etc would need to be extensible since they vary greatly between platforms
2. Autocomplete hint etc would also need to be extensible since reserve words etc are different
3. More importantly - multi-instances - if we are to have one JDBC interpreter we need to have a way of supporting multiple instances of the interpreter? Otherwise we are limiting to work with one JDBC host at a time and can't talk to Hive, PostgresQL, Phoenix etc at the same time.

**Comment by jongyoul** [comment]
@felixcheung 

1. http://docs.oracle.com/javase/7/docs/api/java/sql/DriverManager.html#getConnection(java.lang.String,%20java.util.Properties) would be a good candidate
2. We need to talk about it. Can Zeppelin show the hint from reading a specific setting from Interpreter tab?
3. I'm agree for supporting multiple instances of same Interpreters. Do you have any idea about it? Basically, generic JDBC driver will replace all JDBC drivers if we can use multiple instance.

**Comment by jiekechoo** [comment]
How to get this function?

**Comment by jongyoul** [comment]
@jiekechoo You can find a document on 
http://zeppelin.incubator.apache.org/docs/0.5.5-incubating/interpreter/postgresql.html

1. Set the driver name as you want to `postgresql.driver.name`.
2. Set other properties
3. Save

**Comment by jiekechoo** [comment]
@jongyoul I mean how to use jdbc interpreter.

**Comment by jongyoul** [comment]
@jiekechoo This interpreter is based on jdbc connecttion, and this PR makes PostgreSqlInterpreter change JdbcInterpreter only. Thus you can use PostgrsSqlInterpreter to use Jdbc. @vgmartinez Could you please add docs for this interpreter? If you don't, I'll merge it first add docs later.

**Comment by beeva-victorgarcia** [comment]
Hi @jongyoul,
Sorry for my delay ...;) as we move forward with the discussion about the interpreter jdbc ...? I can improve this withdrawal request with previous recommendations, but I still see a problem with one of the points commented by felixcheung, for example, to have multiple instances of the same artist ... How can we do ...?

**Comment by vgmartinez** [comment]
Hi @jongyoul,
Sorry for my delay ...;) as we move forward with the discussion about the interpreter jdbc ...? I can improve this withdrawal request with previous recommendations, but I still see a problem with one of the points commented by @felixcheung , for example, to have multiple instances of the same artist ... How can we do ...?

**Comment by jongyoul** [comment]
@vgmartinez Sorry for delay. I've committed #455 #517 for supporting multiple instances - connections - Because that's based on this interpreter, it would be easy to fit those features in this interpreter. Could you please check those PRs? And I also hope you contribute the documentations.

**Comment by jongyoul** [comment]
And could you please check https://www.mail-archive.com/dev@zeppelin.incubator.apache.org/msg05504.html at dev@. I, strongly, think every contribution should be tracked.

**Comment by vrajat** [comment]
FWIW - I vote for this implementation of JDBC Interpreter. I understand that your focus is on well-known databases - mysql, postgres etc. 
I am working with a relatively unknown JDBC driver: https://github.com/qubole/quark. This design allows me to use zeppelin with unknown or custom drivers too. I just have to put the jar in the classpath (which is a bit confusing). I have picked up this pull request and updated it to work with master of zeppelin in this branch: https://github.com/vrajat/incubator-zeppelin/tree/JDBCInterpreter.

Please feel free to use my branch if you choose this implementation and give credit to @vgmartinez . He did all the hard work.

**Comment by jongyoul** [comment]
@vrajat Thanks for taking an interest in this issue and being willing to contribute your code. I have a plan to merge several jdbc-like interpreters until 0.6.0-incubating. For that, I'm trying to find PRs related JDBC and contacting contributors to track their contribution. Now, I'm trying to contact the contributor of #211 and this PR is next candidate if I'm not contacting that contributor. And I also think this implementation looks the best, but I think the former contribution should be tracked and accepted at first if that contributor is still willing to give his/her time to contribute. I hope you understand my thoughts and plans. Finally, I will merge additional features of #455 and #517 into jdbc implementation. Please wait a moment and Thank you again for concerning about this issue.

**Comment by jongyoul** [comment]
@vrajat And I also think JDBC interpreter supports all kind of JDBC drivers.

**Comment by vrajat** [comment]
I read your mail. I understand why you are following this process.
I am voting for a requirement that seems to be missing in others but is
satisfied in this one. Just want to make sure you hear all requests.
Thanks for shepherding the JDBC feature. Will wait for it in the main
branch. I'll help in anyway possible.

**Comment by jongyoul** [comment]
@vgmartinez It took a long time to reach this PR. I hope you could share your time to complete this PR. Basically, this implementation looks good to me, but we need to update docs to describe how to use custom Jdbc drivers. Do you have a time to do this?

**Comment by beeva-victorgarcia** [comment]
Hi @jongyoul,
Yes, this week I can complete the PR...

**Comment by jongyoul** [comment]
@vgmartinez Thanks for quick response. And first of all, could you rebase this PR? this PR won't be merged into current master until you rebase it.

**Comment by vgmartinez** [comment]
Hi @jongyoul,
I think it is not necessary to add another interpreter, I think you can improve the interpreter hive to add the jars to the classpath and use the hive interpreter for all jdbc. What do you think?

I tried your implementation in this [branch](https://github.com/vgmartinez/incubator-zeppelin/tree/jdbc_interpreter) and work for:
- postgres
- mysql
- mariadb
- redshift

I do not see useful to have another interpreter where the code is almost the same as the hive.

**Comment by jongyoul** [comment]
@vgmartinez Yes, HiveInterpreter is based on this interpreter, but I thought it's a natural way of you to commit this PR, then I fix to support to multiple connections. It's a way to conserve your contribution and authority. After that, I'll deprecate HiveInterpreter and any other JDBC-like interpreters, and guide your interpreter by using JDBC connections. As you think this is redundant or don't need to conserve authority, I'll change HiveInterpreter to JdbcInterpreter and deprecate any other interpreters. What about your opinion?

**Comment by jongyoul** [comment]
And... Could you please rebase your changes only and ``push --force`` ? That makes commits clear

**Comment by vgmartinez** [comment]
Hi @jongyoul,
Is rebase now...let me know if you need help with testing or anything else.

File: pom.xml [hunk 2]
```diff
@@ -93,7 +93,8 @@
     <module>shell</module>
     <module>hive</module>
     <module>phoenix</module>
-    <module>postgresql</module>
+    <module>geode</module>
```
Comment (jongyoul) [inline pom.xml]: `geode` is not included by default. Could you please remove this?

File: bin/zeppelin-daemon.sh [hunk 3]
```diff
@@ -67,6 +67,12 @@ if [[ -d "${ZEPPELIN_HOME}/zeppelin-server/target/classes" ]]; then
   ZEPPELIN_CLASSPATH+=":${ZEPPELIN_HOME}/zeppelin-server/target/classes"
 fi
 
+# Add jdbc connector jar
```
Comment (jongyoul) [inline bin/zeppelin-daemon.sh]: This looks an example or tutorial, thus could you please remove the below and update docs?

File: jdbc/pom.xml [hunk 4]
```diff
@@ -27,16 +27,17 @@
...
+  <artifactId>zeppelin-jdbc</artifactId>
   <packaging>jar</packaging>
   <version>0.6.0-incubating-SNAPSHOT</version>
-  <name>Zeppelin: PostgreSQL interpreter</name>
+  <name>Zeppelin: JDBC interpreter</name>
   <url>http://www.apache.org</url>
 
   <properties>
-    <postgresql.version>9.4-1201-jdbc41</postgresql.version>
+    <hive.hive.version>0.14.0</hive.hive.version>
```
Comment (jongyoul) [inline jdbc/pom.xml]: And also please remove these two lines.

File: jdbc/pom.xml [hunk 5]
```diff
@@ -54,18 +55,24 @@
...
-      <version>${postgresql.version}</version>
-    </dependency>
-
+	
     <dependency>
       <groupId>com.google.guava</groupId>
       <artifactId>guava</artifactId>
     </dependency>
-
+
```
Comment (jongyoul) [inline jdbc/pom.xml]: Please remove these two dependencies with same reason

**Comment by jongyoul** [comment]
@vgmartinez Thanks for making it clear. You look like that you want to includes some jdbc connectors but it's not appropriate because we don't decide which connectors are supported. Which connector do you think it's suitable for a default? I think postgresql is a good candidate because there is no license and this is based on that. @Leemoonsoo Do you have any recommendation on a default jdbc connector? I propose we decide nothing for now, and describe how to add specific jdbc interpreter on a document. Except that, It looks good to me. Thanks again, @vgmartinez.

**Comment by Leemoonsoo** [comment]
I think postgresql is good candidate to be default, too. Leaving it undecided also good to me.

**Comment by vgmartinez** [comment]
I removed the dependencies and add the postgres dependency  to have it by default without having to add the jar to the classpath. What do you think?

**Comment by jongyoul** [comment]
LGTM. Merging if there isn't no more discussion.

**Comment by vgmartinez** [comment]
Hi @jongyoul @Leemoonsoo,
There is something more to fix or to make the merge...

**Comment by jongyoul** [comment]
@vgmartinez Sorry for late reply. @Leemoonsoo Do you introduce or use the postgresql interpreter from master branch on the upcoming meetup at 21st Jan? If you do, I'll merge this PR after that. Unless you do, I'll merge it right now.

**Comment by Leemoonsoo** [comment]
@jongyoul Thanks for consideration. Please feel free to merge regardless of upcoming meetup. If i need, i can always use specific commit from git repository.

**Comment by jongyoul** [comment]
@Leemoonsoo Thanks for the reply. I'll merge this PR after resolving conflict.

**Comment by jongyoul** [comment]
@vgmartinez Hi, Could you please rebase this PR one more? Some logging codes are added by another PR. After that rebase, I'll merge it immediately.

**Comment by vgmartinez** [comment]
hi @jongyoul,
It ready...in other PR add docs for the interpreter....

**Comment by jongyoul** [comment]
@vgmartinez Thanks, I'll merge it. Did you make a PR for docs? or will you make it?

**Comment by tzolov** [comment]
@vgmartinez , @felixcheung , @asfgit, @Leemoonsoo 
Hello guys, I am sorry to jump a bit late. 
First let me say that I am glad that you have decided to base the generic JDBC interpreter on the work i've done on the PostgreSqlInterpter.
I believe though that the generic interpreter should extend rather than degradating the existing functionality! For example the new generic JDBC intr. hasn't included the SqlCompleter (not sure why since the later is based on JDBC metdata functionality?) and in the same time the last commit you has removed the original PostgreSql Interpreter?  (Actually one single class is still left in. Looks like a slappy commit?)

IMHO This is unacceptable functionality degradation! I would propose to return the PostgreSQL interpreter until JDBC matures? I will try to prepare a hot-fix asap.

**Comment by vgmartinez** [comment]
Hi @tzolov, 
Yes, the interpreter is based on the PostgreSQL was mentioned https://github.com/apache/incubator-zeppelin/pull/608.  I think is that we should work on this interpreter and add auto-completion functionality already available in Postgres and to all others that will be added. what do you think...?

**Comment by tzolov** [comment]
Thanks for the quick response @vgmartinez , 

The SqlCompleter (removed by the last commit) is meant to be 100% JDBC API compliant and hopefully easy to reuse in a generic jdbc interpreter. 

IMO though it is unacceptable to remove a working interpreters until the JDBC interpreter is completed. Some Zeppelin users (like me) use the master to build Zeppelin and now suddenly some essential (for us) functionality is gone. 
I think that as a percussion it has to be confirmed that the new JDBC driver can support an existing (jdbc based) interpreter before the later is removed. 

I suggest returning the PSQL interpreter code ASAP. Let me know if you need help.

In the meantime we can continue working on the generic JDBC. I will help porting the Completer functionality and missing tests.

**Comment by jongyoul** [comment]
@tzolov Hi, your opinion makes sense. @Leemoonsoo also worried about changing Postgresql to Jdbc, but I don't think it's a big problem because I will add auto-completion feature before we release Zeppelin 0.6.0. But I agree that we should also consider those who use Zeppelin from master. I propose the way to support Postgresql and it's completer is to make another PR asap. But we should make sure that this is temporary until we completed JDBC interpreter. I'll restore Postgresql interpreter soon. And personally, sorry for confusing you.

**Comment by jongyoul** [comment]
@vgmartinez Our goal is to move to the generic jdbc interpreter, finally. Could you please contribute a  document?

**Comment by beeva-victorgarcia** [comment]
ok, no problem when I can add docs.

@jongyoul you move the generic jdbc and restore the PostgresQL Interprete?

**Comment by jongyoul** [comment]
@beeva-victorgarcia I keep this PR merged and add Postgresql interpreter again. I think this is more easier way to add Postgresql interpreter.

**Comment by beeva-victorgarcia** [comment]
ok...make sense

**Comment by jongyoul** [comment]
@tzolov Oops. I've found what's the problem. We just delete sqlCompleter only. It's my mistake. I'll fix asap.

**Comment by woonsan** [comment]
Big applause to you guys!
+1 on moving to jdbc interpreter which may also allow other db specific interpreter extensions in the future.
Cheers!

**Comment by tzolov** [comment]
@jongyoul, 
Imo the safest and cleanest solution is be to revert the last (jdbc) commit and to do the changes right before committing the again. 
Then it might be better to create the jdbc interpreter in parallel, and gradually replace the other jdbc-based interpreters once you can prove you don't break existing functionality. 
What do you think?

**Comment by tzolov** [comment]
Same principles that apply for extending/evolving APIs (or Interfaces) should apply here. One shouldn't drop a new incomplete and undocumented API while removing the existing one.

**Comment by jongyoul** [comment]
@tzolov Yes, It would be safe in general case, but in this case, because we drop postgresql interpreter only, I think it's ok that I just add postgresql interpreter. What do you think of it? And I agree what you told is a general rule.

**Comment by jongyoul** [comment]
I fixed my mistakes. Feel free to talk to me if there's some problems. Thanks everyone for taking care of it.

**Comment by beeva-victorgarcia** [comment]
In the next PR I add docs and auto-completion function...sorry for this confusion

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+      }
+    }
+    if (null == connection) {
+      Properties properties = propertiesMap.get(propertyKey);
+      logger.info(properties.getProperty(DRIVER_KEY));
+      Class.forName(properties.getProperty(DRIVER_KEY));
+      String url = properties.getProperty(URL_KEY);
+      String user = properties.getProperty(USER_KEY);
+      String password = properties.getProperty(PASSWORD_KEY);
+      if (null != user && null != password) {
```
Comment (wfhartford) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: I'm curious about this if/else. Is there a specific issue this is working around? JDBC drivers interpret the `user` and `password` properties as the username and password for the connection, thus this condition seems unnecessary. What's more, it prevents passing of arbitrary additional properties along with the connection. Many JDBC drivers can be configured a great deal by adding vendor specific properties when the connection is created. This condition makes that impossible as long as a username and password are provided.

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+        connection = DriverManager.getConnection(url, user, password);
+      } else {
+        connection = DriverManager.getConnection(url, properties);
+      }
```
Comment (wfhartford) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: I'm curious about this if/else. Is there a specific issue this is working around? JDBC drivers interpret the user and password properties as the username and password for the connection, thus this condition seems unnecessary. What's more, it prevents passing of arbitrary additional properties along with the connection. Many JDBC drivers can be configured a great deal by adding vendor specific properties when the connection is created. This condition makes that impossible as long as a username and password are provided.

Comment (jongyoul) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: @whisperstream Hi, sorry for late reply. I agree with you and think it's my mistake and we have to fix that case that you told the above. Do you mind sharing your time to fix it?

**Pull request closed without merging**